### PR TITLE
Instrumentation

### PR DESCRIPTION
**Pull request by twhaples**

### CODE REVIEW

File: spec/creditsafe/client_spec.rb [hunk 1]
```diff
@@ -47,18 +47,11 @@
...
-
       it 'gives a useful error message' do
-        begin
-          method_call
-        rescue Creditsafe::HttpError => err
-          expect(err.message).to include 'Excon::Errors::Timeout'
-        end
+        expect { method_call }.to raise_error(
+          Creditsafe::HttpError,
+          /Excon::Error(?:s)?::Timeout/
```
Comment (twhaples) [inline spec/creditsafe/client_spec.rb]: It used to throw an `Excon::Errors::Timeout` but now throws an `Excon::Error::Timeout`.

File: .rubocop.yml [hunk 2]
```diff
@@ -21,4 +21,4 @@ Metrics/MethodLength:
   Max: 15
 
 Metrics/ClassLength:
-  Max: 150
+  Max: 154
```
Comment (twhaples) [inline .rubocop.yml]: If it's worth refactoring this class to make it shorter I'd rather do it separately

File: .rubocop.yml [hunk 3]
```diff
@@ -21,4 +21,4 @@ Metrics/MethodLength:
   Max: 15
 
 Metrics/ClassLength:
-  Max: 150
+  Max: 152
```
Comment (twhaples) [inline .rubocop.yml]: if there's a refactoring here we should do it separately

File: lib/creditsafe/client.rb [hunk 4]
```diff
@@ -127,15 +127,29 @@ def handle_message_for_response(response)
...
-    rescue => error
-      handle_error(error)
+      notification_payload[:response] = response.body
+    rescue => raw_error
+      begin
+        handle_error(raw_error)
+      rescue => processed_error
+        notification_payload[:error] = processed_error
+        raise processed_error
+      end
```
Comment (twhaples) [inline lib/creditsafe/client.rb]: the error ping-pong going on here is quite silly, but I guess it's a little better than case statement shenanigans in handle_error? maybe?

File: spec/creditsafe/client_spec.rb [hunk 5]
```diff
@@ -1,13 +1,40 @@
...
 RSpec.describe(Creditsafe::Client) do
+  notifications = []
   let(:username) { "b" }
   let(:password) { "c" }
+  before(:all) do
+    ActiveSupport::Notifications.subscribe do |*args|
+      notifications << ActiveSupport::Notifications::Event.new(*args)
+    end
+  end
+  before(:each) { notifications = [] }
```
Comment (twhaples) [inline spec/creditsafe/client_spec.rb]: Unfortunately the execution context in `subscribe` doesn't let me put these in an `@notifications` variable or anything like that. (The rest is the canonical `ActiveSupport::Notifications` collect-all-events recipe).

File: spec/creditsafe/client_spec.rb [hunk 6]
```diff
@@ -157,6 +199,27 @@
...
+        find_company
+        expect(notifications).to match([have_attributes(
+          payload: {
+            request: be_truthy,
+            response: {
+              find_companies_response: include(
+                find_companies_result: include(
+                  messages: {
+                    message: include(
+                      "There are no results matching specified criteria."
```
Comment (Baltoli) [inline spec/creditsafe/client_spec.rb]: I take it that this huge nested hash is just an unfortunate consequence of using SOAP, and that there isn't a lot that could be done to reduce the size

**Comment by Baltoli** [comment]
👍 after `handle_error` etc. is refactored to be clearer on the intent

File: spec/creditsafe/client_spec.rb (hunk 6 above)
Comment (twhaples) [inline spec/creditsafe/client_spec.rb]: Well, we could care about the contents less... but that's pretty much the case yes

**Pull request merged**